projects/slap_manipulation/src/slap_manipulation/dataloaders/peract_loader.py

```diff
@@ -0,0 +1,582 @@
+    def _get_action(
+        self,
+        sample: dict,
+        rotation_resolution: int,
+        idx: int = -1,
+        max_keypoints=6,
+    ):
+        if idx == -1:
+            quat = sample["ee_keyframe_ori_quat"]
+            attention_coordinate = sample["ee_keyframe_pos"]
+            grip = float(sample["proprio"][-1])
+            D = len(self.voxel_sizes)
+            gripper_width = sample["gripper_width_array"]
+        else:
+            quat = sample["ee_keyframe_ori_quat"][idx]
+            attention_coordinate = sample["ee_keyframe_pos"][idx]
+            grip = float(sample["all_proprio"][idx, -1])
+            # gripper_state = sample["target_gripper_state"][idx]
+            # grip = float(sample["proprio"][idx][-1])
+            gripper_width = sample["gripper_width_array"][idx]
+            time_index = float((2.0 * idx) / max_keypoints)
+        quat = utils.normalize_quaternion(quat)
+        if quat[-1] < 0:
+            quat = -quat
+        disc_rot = utils.quaternion_to_discrete_euler(quat, rotation_resolution)
+        trans_indices, attention_coordinates = [], []
+        bounds = np.array(self.scene_bounds)
+        for depth, vox_size in enumerate(
+            self.voxel_sizes
+        ):  # only single voxelization-level is used in PerAct
+            # all of the following args should be numpy
+            index = utils.point_to_voxel_index(
+                attention_coordinate.numpy(), vox_size, bounds
+            )
+            trans_indices.extend(index.tolist())
+            res = (bounds[3:] - bounds[:3]) / vox_size
+            attention_coordinate = bounds[:3] + res * index
+            attention_coordinates.append(attention_coordinate)
+        rot_and_grip_indices = disc_rot.tolist()
+        rot_and_grip_indices.extend([grip])
+        ignore_collisions = [int(0.0)]  # hard-code
+        return (
+            torch.Tensor([trans_indices]),
+            torch.Tensor([rot_and_grip_indices]),
+            torch.Tensor([ignore_collisions]),
+            torch.cat((gripper_width, torch.FloatTensor([grip, time_index]))),
+            torch.Tensor(attention_coordinate).unsqueeze(0),
+        )
+
+    # extract CLIP language features for goal string
+    def _clip_encode_text(self, text):
+        x = self.clip_model.token_embedding(text).type(
+            self.clip_model.dtype
+        )  # [batch_size, n_ctx, d_model]
+
+        x = x + self.clip_model.positional_embedding.type(self.clip_model.dtype)
+        x = x.permute(1, 0, 2)  # NLD -> LND
+        x = self.clip_model.transformer(x)
+        x = x.permute(1, 0, 2)  # LND -> NLD
+        x = self.clip_model.ln_final(x).type(self.clip_model.dtype)
+
+        emb = x.clone()
+        x = (
+            x[torch.arange(x.shape[0]), text.argmax(dim=-1)]
+            @ self.clip_model.text_projection
+        )
+
+        return x, emb
+
+    def clip_encode_text(self, text):
+        """encode text as a sequence"""
```

Comment:
as a sequence using the CLIP text transformer to get a set of per token ebeddings?